src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/yflow/update/action/OnReceivedRemoveResponseAction.java

```diff
@@ -0,0 +1,90 @@
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.flowhs.fsm.yflow.update.action;
+
+import static java.lang.String.format;
+
+import org.openkilda.floodlight.api.response.SpeakerResponse;
+import org.openkilda.floodlight.flow.response.FlowErrorResponse;
+import org.openkilda.wfm.topology.flowhs.fsm.common.actions.HistoryRecordingAction;
+import org.openkilda.wfm.topology.flowhs.fsm.yflow.update.YFlowUpdateContext;
+import org.openkilda.wfm.topology.flowhs.fsm.yflow.update.YFlowUpdateFsm;
+import org.openkilda.wfm.topology.flowhs.fsm.yflow.update.YFlowUpdateFsm.Event;
+import org.openkilda.wfm.topology.flowhs.fsm.yflow.update.YFlowUpdateFsm.State;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.UUID;
+
+@Slf4j
+public class OnReceivedRemoveResponseAction extends
+        HistoryRecordingAction<YFlowUpdateFsm, State, Event, YFlowUpdateContext> {
+    private static final String FAILED_TO_REMOVE_RULE_ACTION = "Failed to remove rule";
+
+    private final int speakerCommandRetriesLimit;
+
+    public OnReceivedRemoveResponseAction(int speakerCommandRetriesLimit) {
+        this.speakerCommandRetriesLimit = speakerCommandRetriesLimit;
+    }
+
+    @Override
+    public void perform(State from, State to, Event event, YFlowUpdateContext context, YFlowUpdateFsm stateMachine) {
+        SpeakerResponse response = context.getSpeakerResponse();
+        UUID commandId = response.getCommandId();
+        if (!stateMachine.hasPendingCommand(commandId)) {
+            log.info("Received a response for unexpected command: {}", response);
+            return;
+        }
+
+        if (response.isSuccess()) {
+            stateMachine.removePendingCommand(commandId);
+            stateMachine.saveActionToHistory("Rule was deleted",
+                    format("The rule was removed: switch %s", response.getSwitchId()));
+        } else {
+            FlowErrorResponse errorResponse = (FlowErrorResponse) response;
+
+            int attempt = stateMachine.doRetryForCommand(commandId);
+            if (attempt <= speakerCommandRetriesLimit) {
+                stateMachine.saveErrorToHistory(FAILED_TO_REMOVE_RULE_ACTION,
+                        format("Failed to remove the rule: commandId %s, switch %s. Error %s. Retrying (attempt %d)",
+                                commandId, errorResponse.getSwitchId(), errorResponse, attempt));
+
+                //TODO: stateMachine.getCarrier().sendSpeakerRequest(removeCommand.makeRemoveRequest(commandId));
+            } else {
+                stateMachine.removePendingCommand(commandId);
```

Comment:
I've changed the order of operations in YFlowCreateFSM to the following: 1. addFailedCommand and then 2. removePendingCommand. Please, check and align.

Reply:
Fixed.